### ISSUE TITLE
False negative for operations on NotImplemented

### ISSUE THREAD

**Issue by 5j9** [microsoft/pyright#9079]
**Describe the bug**
In short, `1 + NotImplemented` does not trigger any type error. This, and the fact that some special methods return `NotImplemented` as return value, may cause some bugs to be left undetected. 

For example, the following code raises `TypeError` at runtime but passes type-checking.
```python
from typing import reveal_type


class MyClass:
    def __add__(self, other):
        if isinstance(other, MyClass):
            return 42
        else:
            return NotImplemented


# Type of x is _NotImplementedType | Literal[42]
# This line will raise TypeError at runtime.
x = MyClass() + 10
y = x / 10
reveal_type(x)
```

```bash
$pyright scratch.py
WARNING: there is a new pyright version available (v1.1.382 -> v1.1.382.post0).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

.../scratch.py
  .../scratch.py:16:13 - information: Type of "x" is "_NotImplementedType | Literal[42]"
0 errors, 0 warnings, 1 information
```
```bash
$ python scratch.py
Traceback (most recent call last):
  File ".../scratch.py", line 14, in <module>
    x = MyClass() + 10
        ~~~~~~~~~~^~~~
TypeError: unsupported operand type(s) for +: 'MyClass' and 'int'
```
```bash
$ pyright --version
WARNING: there is a new pyright version available (v1.1.382 -> v1.1.382.post0).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

pyright 1.1.382
```

**Comment by erictraut** [comment]
Pyright is working correctly here. The typeshed stubs define `NotImplemented` [here](https://github.com/python/typeshed/blob/a94c927642eda093db226eb3e21c3a4681577f85/stdlib/builtins.pyi#L1261). As you can see, it's effectively defined as `Any`, which means that type checkers will not generate an error. This is intentional because of the way `NotImplemented` is used in Python operator overload methods. 

If you think this behavior is undesirable, you could start a discussion with the typeshed maintainers to see if they'd be interested in changing the definition. I'm guessing that there will be reluctance to do so because it would create a lot of churn (including many false positives) in existing typed code bases.

In any case, pyright is working correctly here, so I'm going to close the issue.
